### ISSUE TITLE
[Feature Request] Extend `vi.fn`

### ISSUE THREAD

**Issue by sheremet-va** [vitest-dev/vitest#756]
### Clear and concise description of the problem

We previously had a disscussion that it would be usefull to have a method on `vi.fn` that will defined behaviour when the function is called with this arguments.

Also does anyone else think that `vi.fn` API is too verbose? Maybe we can tighten it a bit. Would love to see suggestions and thoughts.

### Suggested solution

 For example:

Instead of this
```ts
vi.fn((arg1) => {
  if(arg1 === '3') return true
  if(arg1 === '4') return false
  return null
})
```

We can wright something like this (API is not finalized):

```ts
vi.fn()
  .mockReturn(null)
  .on('3').mockReturn(true)
  .on('4').mockReturn(false)
```

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by patak-dev** [comment]
I like the idea of `mock.verify()`, or `verifyMock(mock)`

About an API to define the mapping, maybe we should leave that for user-land? Because your example could be a stand alone utility that generates the implementation:
```js
vi.fn(
  mon().mockReturn(null)
  .on('3').mockReturn(true)
  .on('4').mockReturn(false)
)
```
Or it could be based on an array
```js
vi.fn(mockFrom([
  { return: null },
  { when: '3', return: true },
  { when: '4', return: false }
])
```

Maybe when we get pattern matching in JS users could reach for it https://github.com/tc39/proposal-pattern-matching? I kind of like the simplicity of a function and letting the user do whatever they want inside here.

**Comment by Demivan** [comment]
+1 for making it a separate library. It would be useful with other test runners too.

By the way, there it a `MoqJS` library: https://github.com/slavik57/moqjs. I did not think it would be possible to implement it in js land.

**Comment by sheremet-va** [comment]
-1 for a separate package
I think we can give better DX out of the box instead of outdated jest model. Having it is good for easier migration, but I think it lacks a little bit with its API

**Comment by rogatty** [comment]
I found this issue when looking for a Vitest based alternative for https://github.com/timkindberg/jest-when. I guess I have to migrate to a list of ifs, but hopefully this library can serve as an inspiration.

**Comment by travi** [comment]
i'm also looking for a vitest alternative for `jest-when` or sinon's `.withArgs`. use of conditional stubs like this helps to encourage state based testing better since separate call verification can encourage unwanted side-effect based behavior. 

an api like `jest-when` can greatly simplify the complexity that grows quickly when implementing similar conditional behavior inside of `.mockImplementation`. lack of this ability is the main hesitation i currently have when considering moving tests using either sinon with `.withArgs` or jest tests using `jest-when` over to viteset.

**Comment by sheremet-va** [comment]
> I guess I have to migrate to a list of ifs

> i currently have when considering moving tests using either sinon with `.withArgs` or jest tests using `jest-when` over to viteset

I'm pretty sure you can use `jest-when` with Vitest.

**Comment by travi** [comment]
> I'm pretty sure you can use `jest-when` with Vitest. Vitest was design to be used with libraries in jest ecosystem.

i hoped for the same the last time i tried, but ran into errors when attempting to incorporate it. it has been a bit since i've attempted, so i'll try it again and bring back some detail of the error specifics.

**Comment by sheremet-va** [comment]
> i hoped for the same the last time i tried, but ran into errors when attempting to incorporate it. it has been a bit since i've attempted, so i'll try it again and bring back some detail of the error specifics.

Yeah, looking at the source code you might have errors. #1956 is meant to fix this, but we haven't released new version yet

**Comment by travi** [comment]
> #1956 is meant to fix this, but we haven't released new version yet

great to hear! i'll watch for the next release and give it a shot. thanks!

**Comment by travi** [comment]
with [v0.23.0](https://github.com/vitest-dev/vitest/releases/tag/v0.23.0) available, i was able to use `jest-when` with one of my simple projects: https://github.com/form8ion/vitest/commit/1dc288be0cf2266edce054d405eb750fbc81da8e

i did have to enable [globals](https://vitest.dev/config/#globals) for it to work, but seems to work as expected after that. thanks!

**Comment by MWhite-22** [comment]
Replying to this issue as we are in the process of refactoring a jest codebase into vitest.

One of the issues we found is jest-when is used everywhere throughout our test files, which we don't want to refactor out yet, but `@types/jest-when` is overwriting the global vitest types with their jest equivalents (describe, test, expect, etc.)

Temporary fix is patching the `@types/jest-when`, but given that our developers like the functionality of `when(x).calledWith`, it would be great if the vitest mock fns had that either built in, or a type safe version could be added to extend `vi.fn`s. Without relying on (or being named) jest.

**Comment by MWhite-22** [comment]
Just kidding on the above, upgrading to vitest@0.29.2 has broken jest-when.

**Comment by Kobee1203** [comment]
There is the following library: https://github.com/mcous/vitest-when 

It would be great if vitest could directly integrate these functions.

**Comment by nstepien** [comment]
I spy on console warnings/errors to ensure tests are clean and have no issues, but I also want to ignore specific errors like the one from ResizeObserver.

Here's the implementation I have:
```ts
beforeEach(({ onTestFinished }) => {
  vi.spyOn(console, 'warn').mockName('console.warn');

  const errorMock = vi.fn(console.error).mockName('console.error');
  vi.spyOn(console, 'error').mockImplementation(function error(...params) {
    if (
      params[0] instanceof Error &&
      params[0].message === 'ResizeObserver loop completed with undelivered notifications.'
    ) {
      return;
    }

    return errorMock(...params);
  });

  onTestFinished(({ expect }) => {
    expect
      .soft(
        console.warn,
        'console.warn() was called during the test; please resolve unexpected warnings'
      )
      .toHaveBeenCalledTimes(0);
    expect
      .soft(
        errorMock,
        'console.error() was called during the test; please resolve unexpected errors'
      )
      .toHaveBeenCalledTimes(0);
  });
});
```

I have to split the `console.error` spy in two to avoid increasing the call count when ignoring the ResizeObserver error, and to call the original `console.error` implementation.

I wonder if a `Mock.filter` or `Mock.map` API would work?

For example:
```ts
// ignore filtered out calls
// return true  -> increase call count, call underlying implementation (original or mocked)
// return false -> do not increase call count, do not call underlying implementation
vi.spyOn(console, 'error').filter((...params) => {
  if (
    params[0] instanceof Error &&
    params[0].message === 'ResizeObserver loop completed with undelivered notifications.'
  ) {
    return false;
  }

  return true;
});

// mapped implementation
vi.spyOn(console, 'error').map((cb, ...params) => {
  if (condition) {
    // custom handling / implementation
    return;
  }

  // call underlying implementation (original or mocked)
  return cb(...params);
});
```

Chaining could possibly make sense:

```ts
vi.spyOn(console, 'error').filter(...).mockImplementation(...);
```

**Comment by sheremet-va** [comment]
The team decided to adopt this API:

```ts
const spy = vi.fn()

const when = vi.when(spy)
  .calledWith(3).thenThrow()
  .calledWith(3).thenReturnOnce(3)
```

There will be a matcher to check that all behaviours were called:

```ts
expect(when).toHaveBeenExhausted()
```

`Spy` and `When` are different objects to avoid overriding properties of the original object (like when using with `vi.spyOn`).

The `calledWith` arguments are checked with the same equals used in `toEqual`, so asymmetric matchers are supported automatically. This makes @nstepien's use case solvable:

```ts
// defaults to the original implementation, unless called with an object that has a matching message
vi.when(vi.spyOn(console, 'warn'))
  .calledWith(
    expect.objectContaining({ message: 'ResizeObserver loop completed with undelivered notifications.' })
  )
  .thenReturn(undefined)
```

**Comment by nstepien** [comment]
Does this also control how calls to spies are recorded?
i.e. in your example, if this specific error message gets logged, will it influence `toHaveBeenCalledTimes()` assertions?

**Comment by sheremet-va** [comment]
> Does this also control how calls to spies are recorded? i.e. in your example, if this specific error message gets logged, will it influence `toHaveBeenCalledTimes()` assertions?

Yes, I don't think it's reasonable to ignore any of them since the spy is always applied
